icicle/tests/test_hash_api.cpp

```diff
@@ -315,54 +322,216 @@ class HashSumBackend : public HashBackend
-    const int nof_hashes = input_size / layer_hash.default_input_chunk_size();
+    output_size = input_size_temp * layer_hash.output_size() / layer_hash.default_input_chunk_size();
+    const int nof_hashes = input_size_temp / layer_hash.default_input_chunk_size();
 
     auto config = default_hash_config();
     config.batch = nof_hashes;
     layer_hash.hash(layer_in.get(), layer_hash.default_input_chunk_size(), config, layer_out.get());
 
     // copy output outputs to inputs before moving to the next layer
     memcpy(layer_in.get(), layer_out.get(), output_size);
-    input_size = output_size;
+    input_size_temp = output_size;
   }
 
   // Compare computed root with the tree's root
   auto [root, root_size] = tree.get_merkle_root();
-
   for (int i = 0; i < root_size; i++) {
-    ASSERT_EQ(root[i], layer_out[i]) << "Mismatch in root[" << i << "]";
+    if (root[i] != layer_out[i]) { return false; }
   }
+  return true;
 }
 
+/**
+ * @brief Wrapper to the non-template version of is_valid_tree above, allowing to insert different types of arrays as
+ * inputs. Builds tree in a straight-forward single-threaded manner and compares the result with Icicle's calculation.
+ * @param tree - Merkle tree to test.
+ * @param input_size - Size of input in bytes.
+ * @param inputs - Input as a byte array.
+ * @param hashes - Vector of hashes of each layer in the tree above.
+ * @return True if the tree's calculations (icicle and test) match.
+ */
 template <typename T>
-void assert_valid_tree(
+bool is_valid_tree(
   const MerkleTree& tree,
   int nof_inputs,
   const T* inputs,
   const std::vector<Hash>& hashes,
   const MerkleTreeConfig& config)
 {
-  return assert_valid_tree(tree, nof_inputs * sizeof(T), reinterpret_cast<const std::byte*>(inputs), hashes, config);
+  return is_valid_tree(
+    tree, nof_inputs * sizeof(T), sizeof(T), reinterpret_cast<const std::byte*>(inputs), hashes, config);
+}
+
+/**
+ * @brief Function used by the HashApiTest to test the various Merkle trees defined in the tests below. Checks
+ * validity of the tree construction, and correctness/incorrectness of valid/invalid proofs generated by the tree.
+ * @param hashes - Vector of hashes of each layer in the tree above.
+ * @param config - Merkle tree config (Mostly irrelevant for cpu tests).
+ * @param output_store_min_layer - Store layer parameter for the Merkle tree builder.
+ * @param nof_leaves - Size of the T leaves array.
+ * @param leaves - Aforementioned leaves array.
+ * @param explicit_leaf_size - Optional. Size of each leaf element in case that leaves is given as a byte array. NOTE
+ *                             test will fail if this value isn't default (1) and T != std::byte
+ */
+template <typename T>
+void test_merkle_tree(
+  const std::vector<Hash>& hashes,
+  const MerkleTreeConfig& config,
+  const int output_store_min_layer,
+  int nof_leaves,
+  const T* leaves,
+  unsigned explicit_leaf_size = 1)
```

Comment:
why default value?

Reply:
This value is only needed in cases where the actual leave size is different than sizeof(T) (For example MerkleTreeLarge test) to avoid having to fill it everytime, a default value was set. Given that this value is used multiplicatively inside the function, a value of 1 is set to be the default.